PatternFly list

andresgalante opened the pull request:
## Description
This PR adds a bare bones lists, which is a good nest for list-view
The full description is on the jira story [PTNFLY-2188](https://patternfly.atlassian.net/browse/PTNFLY-2188)

@bleathem will be able to give more details on the Jquery implementation.

## Todos
- [ ] cross browser test
- [ ] Are you sure it works on IE9?
- [ ] Is it reponsive?

##  Link to rawgit
https://rawgit.com/andresgalante/patternfly/pf-list-dist/dist/tests/list-pf.html

srambach commented:
I love having this generic listview!
First impression from just looking at the example, not code yet :-)
Is it intended that the hover only works on top-level items?  
I'm not sure whether the gray background on an expanded first line is just to show that it's the first line, or is it that it's selected? I agree also that the colors should be different to better illustrate the different roles of the highlighting, completely apart from whatever the final visual design is. That will also help to demonstrate clearly what is taking precedence - expansion color, selection, or hover.

andresgalante commented:
@srambach @rhamilto Yeap, that makes a lot of sense, @kybaker Can you please take a look at this?

@rhamilto I will add the arrow as an optional element. I though that since we need to place the arrow on the left in some way (floating, flexing or grid) then that would interfear with how the user wants to build their list.

srambach commented on src/less/variables.less:
```diff
@@ -49,6 +49,11 @@
+@list-pf-active-border-color:                                       @color-pf-black-400;
+@list-pf-header-background-color:                                   @color-pf-black-200;
+@list-pf-hover-background-color:                                    @color-pf-black-200;
+@list-pf-padding:                                                   20px;
```
Use a spacer variable here?

jgiardino commented:
This looks really nice, Andres!

Piggy-backing on Sarah and Rob's comments about hover color, I really like how hover works in the table examples here, where the color doesn't change but just becomes more saturated on hover: http://blog.andresgalante.com/patternfly-atomic/?p=viewall-basics-table

I don't know how much of this work is intended to address the visual styling we currently use, but 
it would be nice to avoid what we have here: https://rawgit.com/patternfly/patternfly/master-dist/dist/tests/list-view-simple-expansion.html
where selected list items are highlighted blue, unless they're expanded. It seems like selection is more important and therefore should override the color highlight associated with expansion. 

Since I've been reviewing the [a11y guidelines](http://a11yproject.com/checklist.html), are you planning to add keyboard and screenreader support to this? Supporting keyboard and screenreader access might be simpler if we always include the arrow, by giving screenreaders and keyboard users something to place focus on.

rhamilto commented on src/less/list-pf.less:
```diff
@@ -0,0 +1,47 @@
+//
+// Close icons
```
Comment needs updating

kybaker commented:
@andresgalante I agree with Robb about removing the shadow. I think if we replaced the hover blue with a lighter gray than the selected state background it would have the same smooth transition that we see in your PF5 data table example.

andresgalante commented:
@rhamilto @kybaker I've removed the shadow and changed the hover color

rhamilto commented:
@andresgalante, ETA on when this will merge?  We're assuming it'll be in v3.21.0 in 2 weeks?

andresgalante commented:
@srambach @jgiardino can we merge this one?

jgiardino commented:
LGTM!

rhamilto commented:
@andresgalante, regarding your most recent push:  I wouldn't expect the hover on a non-clickable/non-header area.

rhamilto commented:
And an already expanded header probably shouldn't hover, either?

srambach commented:
I was going to add that the light hover on a dark, expanded header seemed like it should darken it further, along the lines of the table example the @jgiardino mentions above.
@rhamilto, to you does the hover only indicate expandable then? Not that you can do something (expand or collapse)?

rhamilto commented:
> @rhamilto, to you does the hover only indicate expandable then? Not that you can do something (expand or collapse)?

I'm thinking so.  Because the non-header area will have all different kinds of content in it that may or may not be individually clickable.

andresgalante commented:
:) makes sense, I'll role back that change. 

Does it make sense to add a modifier class to enable or disable hovers in each row or do we just go back to the way it was before?

srambach commented:
I like the idea of hover by default but a modifier class to disable hovers. Or vice-versa but that seemed the more likely case.

rhamilto commented:
> Or vice-versa but that seemed the more likely case.

In Bootstrap, hover behavior on tables is opt-in.  I think it would be good to be consistent with what Bootstrap is doing.

jgiardino commented:
@rhamilto's suggestion to follow bootstrap's model makes sense. But to me, that means one class for `list-pf`. I'm not sure if you were thinking a class for each `list-pf-container` or not when you said this, @andresgalante:

> Does it make sense to add a modifier class to enable or disable hovers in each row

andresgalante commented:
@srambach @jgiardino @rhamilto I've added a `-hover` class, and it displays a bg when the row is clicable.

The hover class on tables adds the background to all rows.

I am a bit lost here, What do we want to do? 😄 thanks!

rhamilto commented:
I think the hover should be present on headers that aren't already expanded by default (with no option to disable).

And hover on non-header (content) areas should be opt-in if we deem necessary.  I personally don't think there is a need to have hover on the content area.

jgiardino commented:
Following up on Rob's comment regarding hover over expanded headers, I could go with either of these solutions:

- no hover color change 
- hover color change to a darker color (but not a lighter color like it is now) e.g.:
```
.list-pf-hover .active > .list-pf-container:not(:last-child):hover {
    background-color: #dfdfdf;
}
```
_Although in this example, I don't think the `:not(:last-child)` is needed? And `#dfdfdf` is just a random darker color for example purposes but I think it's too dark._

I also don't think there is a need to have hover over the content area, but I don't think I would make that optional either. That seems like such an edge case that it would need custom code anyway.  

I don't have a strong opinion on whether the hover over headers should be optional.

andresgalante commented:
ok, lets try this one more time.
I've removed the opt-in hover and added hovers only on non active "headers", like the open shift example.

Does it make more sense now?

rhamilto commented:
I'm still seeing the hover color on "open" headers.  :/

andresgalante commented:
@rhamilto wait a few mins, travis takes time to compile and re deploy

rhamilto commented:
@spadgett and I just reviewed the latest and noticed a couple things:

* the cursor should probably change to pointer (hand) when mousing over the headers.
* currently there is no keyboard navigation.  should there be?
* is the chevron obvious enough?  @spadgett and I are partial to the larger chevron with a "pipe" that we're using in OpenShift.  


 [1]

![screen shot 2017-02-17 at 11 29 36 am](https://cloud.githubusercontent.com/assets/895728/23073726/71ff6d80-f504-11e6-9cf8-85a5f6585a50.PNG)

![screen shot 2017-02-17 at 11 31 10 am](https://cloud.githubusercontent.com/assets/895728/23073794/9de07930-f504-11e6-991e-d98c88b892e2.PNG)

![screen shot 2017-02-17 at 11 31 15 am](https://cloud.githubusercontent.com/assets/895728/23073822/b29c129e-f504-11e6-9f49-b4563e75e875.PNG)

kybaker commented:
@andresgalante I agree with Robb about the pipe and chevron. I know the focus was not on the content in your example, but these elements are a part of the interaction.

andresgalante commented:
Thanks! I've added the pipe and increase the sice of the chevron. Also keyboard navigation now works.

Thanks for the reviews!

rhamilto commented:
The chevrons should be pointed down since these two sections are expanded by default.

![screen shot 2017-02-17 at 4 07 17 pm](https://cloud.githubusercontent.com/assets/895728/23083346/3c4cbea0-f52b-11e6-95d6-c8e48eb4ee94.PNG)

andresgalante commented:
you are right, just sent an update, thanks
